### PR TITLE
Add endpoints to deactivate and restore config

### PR DESCRIPTION
**Pull request by lelenanam**
should call this to deactivate config after instance deletion

### CODE REVIEW

**Comment by lelenanam** [comment]
Fixes https://github.com/weaveworks/service/issues/1595

**Comment by bboreham** [comment]
You need to run `make update-vendor` after `dep` commands - see https://github.com/weaveworks/cortex/blob/master/README.md#vendoring

**Comment by bboreham** [comment]
Now I've looked at this a bit more, it will do something slightly surprising.

The DB has potentially many configs for an instance, each with a new ID number.
`DB.GetConfig()` selects the non-deleted ones, `limit 1 order by id DESC`.

The code in this PR will mark _all_ rows for an instance as deleted on `DeactivateConfig()` and mark _all_ of them as undeleted on `RestoreConfig()`.  So that's ok, but surprising to me.

This wouldn't play well if, for instance, some other component was marking individual rows as deleted.  (Although I can't see anything like that).

Another issue is that `ruler` will only see the config as changed if there is one for the same instance with a higher ID number.

I'm not sure what to do about this. 

#332 covers garbage-collecting old rows

**Comment by bboreham** [comment]
There is an issue to report deleted configs - #329

File: pkg/configs/db/timed.go [hunk 1]
```diff
@@ -69,6 +69,18 @@ func (t timed) GetConfigs(since configs.ID) (cfgs map[string]configs.View, err e
 	return
 }
 
+func (t timed) DeactivateConfig(userID string) (err error) {
+	return t.timeRequest("DeactivateConfigs", func(_ context.Context) error {
```
Comment (rndstr) [inline pkg/configs/db/timed.go]: `"DeactivateConfig"` & same in `traced.go`

File: pkg/configs/db/timed.go [hunk 2]
```diff
@@ -69,6 +69,18 @@ func (t timed) GetConfigs(since configs.ID) (cfgs map[string]configs.View, err e
...
+		return t.d.DeactivateConfig(userID)
+	})
+}
+
+func (t timed) RestoreConfig(userID string) (err error) {
+	return t.timeRequest("RestoreConfigs", func(_ context.Context) error {
```
Comment (rndstr) [inline pkg/configs/db/timed.go]: `"RestoreConfig"` & same in `traced.go`

File: pkg/configs/db/memory/memory.go [hunk 3]
```diff
@@ -32,12 +34,22 @@ func New(_, _ string) (*DB, error) {
 	}, nil
 }
 
+func isActive(c config) bool {
+	if c.deletedAt.IsZero() {
```
Comment (rndstr) [inline pkg/configs/db/memory/memory.go]: `return c.deletedAt.IsZero()`?

File: pkg/configs/db/postgres/postgres.go [hunk 4]
```diff
@@ -42,8 +46,32 @@ type dbProxy interface {
...
+// dbWait waits for database connection to be established
+func dbWait(db *sql.DB) error {
+	deadline := time.Now().Add(dbTimeout)
+	var err error
+	for tries := 0; time.Now().Before(deadline); tries++ {
+		err = db.Ping()
+		if err == nil {
+			return nil
+		}
+		log.Debugf("db connection not established, error: %s; retrying...", err)
```
Comment (rndstr) [inline pkg/configs/db/postgres/postgres.go]: What about using `WARN` here? Unless `DEBUG` is the default level, it doesn't say why it's stalling

**Comment by bboreham** [comment]
In regard to "ruler will only see the config as changed if there is one for the same instance with a higher ID number" - #329, Elena and I discussed IRL an idea to change the way deleted configs are interpreted.
(@lelenanam if you wrote this up somewhere I apologise; I can't find it)

At present, the DB queries fetch the last active (non-deleted) config.  So you can revert from the latest config to the one before by setting `deleted_at` on one row.  We don't know of any code that does this.

The code that updates ruler has a note of the highest ID number it has seen, and polls looking for active configs with higher numbers.  So it will never report a deleted row, new or otherwise.

Suppose we change the behaviour to return the config with the highest ID, deleted or not.  Then we can implement "delete a config" by adding a single new row with `deleted_at` set.  Ruler would be notified of this new row; it would have to be enhanced to tell the difference between active and deleted configs.

If we wanted a 'revert to previous' feature, it would now be implemented by adding a new row with the previous config.

cc @jml @tomwilkie

**Comment by lelenanam** [comment]
Changed deactivation as @bboreham described (add new config row with `deletedAt`).
But we also need extra code to stop receiving alerts.

**Comment by lelenanam** [comment]
@bboreham PTAL

File: pkg/configs/db/postgres/postgres.go [hunk 5]
```diff
@@ -214,17 +240,65 @@ func (d DB) findRulesConfigs(filter squirrel.Sqlizer) (map[string]configs.Versio
...
 func (d DB) GetRulesConfigs(since configs.ID) (map[string]configs.VersionedRulesConfig, error) {
 	return d.findRulesConfigs(squirrel.And{
-		activeConfig,
+		allConfigs,
 		squirrel.Gt{"id": since},
 	})
 }
 
+// GetLastConfig gets a last configuration (active or deleted).
+func (d DB) GetLastConfig(userID string) (configs.View, error) {
```
Comment (bboreham) [inline pkg/configs/db/postgres/postgres.go]: I have a feeling this function isn't necessary: `GetConfig` is good enough for the purpose.

File: pkg/configs/db/postgres/postgres.go [hunk 6]
```diff
@@ -214,17 +240,65 @@ func (d DB) findRulesConfigs(filter squirrel.Sqlizer) (map[string]configs.Versio
...
+		Limit(1).
+		QueryRow().Scan(&cfgView.ID, &cfgBytes)
+	if err != nil {
+		return cfgView, err
+	}
+	err = json.Unmarshal(cfgBytes, &cfgView.Config)
+	return cfgView, err
+}
+
+// SetDeletedAtConfig sets a deletedAt for configuration.
```
Comment (bboreham) [inline pkg/configs/db/postgres/postgres.go]: I think it's worth commenting why "set" is actually an "insert".

**Comment by bboreham** [comment]
What happens in `scheduler.addNewConfigs()` when a deleted record update arrives in the middle of operation?

**Comment by lelenanam** [comment]
> What happens in scheduler.addNewConfigs() when a deleted record update arrives in the middle of operation?

I suppose the same when config changes in the middle of the operation, this config will be updated/deleted on next `scheduler.pollInterval` tick.

**Comment by lelenanam** [comment]
@bboreham PTAL

**Pull request merged**